client/gui/InterfaceObjectConfigurable.cpp

```diff
@@ -0,0 +1,204 @@
+/*
+* InterfaceBuilder.cpp, part of VCMI engine
+*
+* Authors: listed in file AUTHORS in main folder
+*
+* License: GNU General Public License v2.0 or later
+* Full text of license available in license.txt file, in main folder
+*
+*/
+
+#include "StdInc.h"
+
+#include "InterfaceObjectConfigurable.h"
+
+#include "../CGameInfo.h"
+#include "../gui/CAnimation.h"
+#include "../gui/CGuiHandler.h"
+#include "../widgets/CComponent.h"
+#include "../widgets/Buttons.h"
+#include "../widgets/MiscWidgets.h"
+#include "../widgets/ObjectLists.h"
+#include "../widgets/TextControls.h"
+#include "../windows/GUIClasses.h"
+#include "../windows/InfoWindows.h"
+
+#include "../../lib/CGeneralTextHandler.h"
+
+
+InterfaceObjectConfigurable::InterfaceObjectConfigurable(const JsonNode & config):
+	CIntObject()
+{
+	init(config);
+}
+
+InterfaceObjectConfigurable::InterfaceObjectConfigurable():
+	CIntObject()
+{
+}
+
+void InterfaceObjectConfigurable::addCallback(const std::string & callbackName, std::function<void(int)> callback)
+{
+	callbacks[callbackName] = callback;
+}
+
+void InterfaceObjectConfigurable::init(const JsonNode &config)
+{
+	OBJ_CONSTRUCTION;
+	int unnamedObjectId = 0;
+	const std::string unnamedObjectPrefix = "__widget_";
+	
+	for(const auto & item : config["items"].Vector())
+	{
+		std::string name = item["name"].isNull()
+						? unnamedObjectPrefix + std::to_string(unnamedObjectId++)
+						: item["name"].String();
+		widgets[name] = buildWidget(item);
+	}
+}
+
+std::string InterfaceObjectConfigurable::buildText(const JsonNode & config) const
+{
+	if(config.isNull())
+		return "";
+	
+	if(config.isNumber())
+	{
+		return CGI->generaltexth->allTexts[config.Integer()];
+	}
+	return config.String();
+}
+
+std::shared_ptr<CIntObject> InterfaceObjectConfigurable::buildWidget(const JsonNode & config)
+{
+	assert(!config.isNull());
+	auto type = config["type"].String();
+	
+	int x = 0, y = 0;
```

Comment:
Perhaps use `Point` instead of individual x/y?
Also, if you wish - feel free to add constructor to Button/Label/Picture/etc that accepts Point instead of x/y